ipfs-8-to-9/migration/swapper.go

```diff
@@ -0,0 +1,265 @@
+			errored++
+			continue
+		}
+
+		oldKey := ds.NewKey(res.Key)
+		c, err := dsKeyToCid(oldKey)
+		if err != nil {
+			// complain if we find anything that is not a CID but
+			// leave it as it is.
+			log.Log("could not parse %s as a Cid", oldKey)
+			continue
+		}
+		if c.Version() == 0 { // CidV0 are multihashes, leave them.
+			continue
+		}
+
+		// Cid Version > 0
+		mh := c.Hash()
+		newKey := dshelp.MultihashToDsKey(mh)
+		err = sw.swap(oldKey, newKey)
+		if err != nil {
+			log.Error("swapping %s for %s: %s", oldKey, newKey, err)
+			errored++
+			continue
+		}
+	}
+
+	// final sync
+	err := sw.sync()
+	if err != nil {
+		log.Error("error performing last sync: %s", err)
+		errored++
+	}
+
+	return sw.swapped, errored
+}
+
+// unswap worker takes notifications from unswapCh (as they would be sent by
+// the swapWorker) and undoes them. It ignores NotFound errors so that reverts
+// can succeed even if they failed half-way.
+func (cswap *CidSwapper) unswapWorker(unswapCh <-chan Swap) (uint64, uint64) {
+	var errored uint64
+
+	swker := &swapWorker{
+		store:  cswap.Store,
+		swapCh: cswap.SwapCh,
+	}
+
+	// Process keys from the results channel
+	for sw := range unswapCh {
+		err := swker.swap(sw.New, sw.Old)
+		if err == ds.ErrNotFound {
+			log.Log("could not revert %s->%s. Was it already reverted? Ignoring...", sw.Old, sw.New)
+			continue
+		}
+		if err != nil {
+			log.Error("swapping %s for %s: %s", sw.New, sw.Old, err)
+			errored++
+			continue
+		}
+	}
+
+	// final sync
+	err := swker.sync()
+	if err != nil {
+		log.Error("error performing last sync: %s", err)
+		errored++
+	}
+
+	return swker.swapped, errored
+}
+
+// swapWorker swaps old keys for new keys, syncing to disk regularly
+// and notifying swapCh of the changes.
+type swapWorker struct {
+	swapped     uint64
+	curSyncSize uint64
+
+	swapCh chan Swap
+	store  ds.Batching
+
+	toDelete []ds.Key
+}
+
+// swap replaces old keys with new ones. It Syncs() when the
+// number of items written reaches SyncSize. Upon that it proceeds
+// to delete the old items.
+func (sw *swapWorker) swap(old, new ds.Key) error {
+	v, err := sw.store.Get(old)
+	vLen := uint64(len(v))
+	if err != nil {
+		return err
+	}
+	if err := sw.store.Put(new, v); err != nil {
+		return err
+	}
+	sw.toDelete = append(sw.toDelete, old)
+
+	sw.swapped++
+	sw.curSyncSize += vLen
+
+	if sw.swapCh != nil {
+		sw.swapCh <- Swap{Old: old, New: new}
+	}
+
+	// We have copied about 10MB
+	if sw.curSyncSize >= SyncSize {
+		sw.curSyncSize = 0
+		err = sw.sync()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (sw *swapWorker) sync() error {
+	// Sync all the new keys to disk
+	err := sw.store.Sync(ds.NewKey("/"))
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	// Delete all the old keys
+	for _, o := range sw.toDelete {
+		if err := sw.store.Delete(o); err != nil {
+			return err
+		}
+	}
+	sw.toDelete = nil
+
+	// Sync again.
+	err = sw.store.Sync(ds.NewKey("/"))
```

Comment:
Thinking about this a bit, we actually don't need this second sync. We just need adds to persist before we delete, a final sync at the very end will sync the any unfinished deletes.

Reply:
Yeah, I did it here so that I don't have to do a final sync in two places, but I'll change that.